x509/helper.go

```diff
@@ -23,6 +23,19 @@ import (
 	"github.com/notaryproject/notation-core-go/signature"
 )
 
+// validateSelfSignedLeaf validates a self-signed leaf certificate.
+//
+// A self-signed leaf certificate must have the same subject and issuer.
+func validateSelfSignedLeaf(cert *x509.Certificate) error {
```

Comment:
It seems that we already have an `isSelfSigned` helper function at line 39? Can we make use of it?

Reply:
```go
func isSelfSigned(cert *x509.Certificate) (bool, error) {
	return isIssuedBy(cert, cert)
}

func isIssuedBy(subject *x509.Certificate, issuer *x509.Certificate) (bool, error) {
	if err := subject.CheckSignatureFrom(issuer); err != nil {
		return false, err
	}
	return bytes.Equal(issuer.RawSubject, subject.RawIssuer), nil
}
```
We cannot use this function because `CheckSignatureFrom` will enforces `KeyUsageCertSign`, which is required for root cert but not for self-signed leaf cert.

Reply:
I see. Just trying to make concepts clear here. Based on our [spec](https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md#other-requirements), `Valid certificate chain MUST contain a root certificate.` 
In this case, for a valid cert chain of length 1, this only cert is still called `root`. (we're just validating it with the `leaf rules`.)

Therefore, for a `self-signed` cert, I think there's only one `valid` scenario in Notation, i.e. `Root cert of a cert chain`.

So a better design might be combining your new function `validateSelfSignedLeaf` into `isSelfSigned`, adding a flag to distinguish between `cert chain length gt 1` and `cert chain length eq 1`.

Reply:
Updated and consolidated the logic for validating self-signed certificates in the `validateSelfSignedCert` function. Please take a look.